File: companion-android/app/src/main/java/me/fiksu/esp32map/companion/integration/location/HeadingTrail.kt
package me.fiksu.esp32map.companion.integration.location

import me.fiksu.esp32map.companion.domain.CoordinatePoint
import kotlin.math.PI
import kotlin.math.abs
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.sqrt

/**
 * Small ring buffer of recent GPS fixes that derives a smoothed travel
 * heading. Matches the web + iOS + runtime-core contract for spec line 110
 * (authoritative): when the rider is moving, the camera rotates to the
 * GPS-derived direction of travel, overriding the route-segment bearing.
 * Returns `null` while stationary / no usable trail.
 */
class HeadingTrail(
    private val maxAgeMs: Long,
    private val maxFixes: Int,
    private val minDisplacementM: Double,
    private val smoothingAlpha: Double,
) {
    private data class Fix(val point: CoordinatePoint, val timestampMs: Long)

    private val fixes: ArrayDeque<Fix> = ArrayDeque()
    private var smoothed: Double? = null

    fun recordFix(point: CoordinatePoint, timestampMs: Long) {
        evictOld(timestampMs)
        fixes.addLast(Fix(point, timestampMs))
        while (fixes.size > maxFixes) fixes.removeFirst()
        val raw = computeRawHeading() ?: return
        val prev = smoothed
        smoothed = if (prev == null) {
            raw
        } else {
            val delta = shortestSignedDelta(prev, raw)
            normalize360(prev + delta * smoothingAlpha)
        }
    }

    val travelHeadingDegrees: Double?
        get() = smoothed

    fun reset() {
        fixes.clear()
        smoothed = null
    }

    private fun evictOld(nowMs: Long) {
        val cutoff = nowMs - maxAgeMs
        while (fixes.isNotEmpty() && fixes.first().timestampMs < cutoff) fixes.removeFirst()
        if (fixes.isEmpty()) smoothed = null
    }

    private fun computeRawHeading(): Double? {
        if (fixes.size < 2) return null
        val first = fixes.first().point
        val last = fixes.last().point
        val metersPerDegLat = 111_320.0
        val meanLat = (first.latitude + last.latitude) / 2.0 * PI / 180.0
        val dNorth = (last.latitude - first.latitude) * metersPerDegLat
        val dEast = (last.longitude - first.longitude) * cos(meanLat) * metersPerDegLat
        val displacement = sqrt(dNorth * dNorth + dEast * dEast)
        if (displacement < minDisplacementM) return null
        return normalize360(atan2(dEast, dNorth) * 180.0 / PI)
    }

    private fun normalize360(deg: Double): Double {
        val r = deg % 360.0
        return if (r < 0) r + 360.0 else r
    }

    private fun shortestSignedDelta(from: Double, to: Double): Double {
        var d = (to - from + 540.0) % 360.0 - 180.0
        if (d == -180.0) d = 180.0
        return d
    }
}

@Suppress("unused")
private fun Double.absLocal() = abs(this)

File: companion-apps/ios/CompanionApp/Integration/Location/HeadingTrail.swift
import Foundation

/// Small ring buffer of recent GPS fixes that derives a smoothed travel
/// heading. Matches the web + runtime-core contract (spec line 110,
/// authoritative): when the rider is moving, the camera rotates to the
/// GPS-derived direction — this overrides the route-segment bearing.
/// Returns nil while stationary / no usable trail.
@MainActor
final class HeadingTrail {
    private struct Fix {
        let point: CoordinatePoint
        let timestampMs: Int64
    }

    private var fixes: [Fix] = []
    private var smoothedDegrees: Double?

    let maxAgeMs: Int64
    let maxFixes: Int
    let minDisplacementM: Double
    let smoothingAlpha: Double

    init(maxAgeMs: Int64, maxFixes: Int, minDisplacementM: Double, smoothingAlpha: Double) {
        self.maxAgeMs = maxAgeMs
        self.maxFixes = maxFixes
        self.minDisplacementM = minDisplacementM
        self.smoothingAlpha = smoothingAlpha
    }

    func recordFix(_ point: CoordinatePoint, timestampMs: Int64) {
        evictOld(nowMs: timestampMs)
        fixes.append(Fix(point: point, timestampMs: timestampMs))
        if fixes.count > maxFixes { fixes.removeFirst() }
        guard let raw = computeRawHeading() else { return }
        if let prev = smoothedDegrees {
            let delta = shortestSignedDelta(from: prev, to: raw)
            smoothedDegrees = normalize360(prev + delta * smoothingAlpha)
        } else {
            smoothedDegrees = raw
        }
    }

    var travelHeadingDegrees: Double? { smoothedDegrees }

    func reset() {
        fixes.removeAll()
        smoothedDegrees = nil
    }

    private func evictOld(nowMs: Int64) {
        let cutoff = nowMs - maxAgeMs
        while let first = fixes.first, first.timestampMs < cutoff { fixes.removeFirst() }
        if fixes.isEmpty { smoothedDegrees = nil }
    }

    private func computeRawHeading() -> Double? {
        guard fixes.count >= 2 else { return nil }
        let first = fixes[0].point
        let last = fixes[fixes.count - 1].point
        let metersPerDegLat = 111_320.0
        let meanLat = ((first.latitude + last.latitude) / 2.0) * .pi / 180.0
        let dNorth: Double = (last.latitude - first.latitude) * metersPerDegLat
        let dEast: Double = (last.longitude - first.longitude) * cos(meanLat) * metersPerDegLat
        let displacement: Double = (dNorth * dNorth + dEast * dEast).squareRoot()
        if displacement < minDisplacementM { return nil }
        return normalize360(atan2(dEast, dNorth) * 180.0 / .pi)
    }
}

private func normalize360(_ deg: Double) -> Double {
    let r = deg.truncatingRemainder(dividingBy: 360.0)
    return r < 0 ? r + 360.0 : r
}

private func shortestSignedDelta(from a: Double, to b: Double) -> Double {
    var d = (b - a + 540.0).truncatingRemainder(dividingBy: 360.0) - 180.0
    if d == -180.0 { d = 180.0 }
    return d
}
